### PR TITLE
Allow custom client verbs to be generated using client-gen

### PR DESCRIPTION
**Pull request by mfojtik**
This change will allow to define custom verbs for resources using the following new tag:

```
// +genclient:method=Foo,verb=create,subresource=foo,input=Bar,output=k8s.io/pkg/api.Blah
```

This will generate client method `Foo(bar *Bar) (*api.Blah, error)` (format depends on the particular verb type)

With this change we can add `UpdateScale()` and `GetScale()` into all scalable resources. Note that intention of this PR is not to fix the Scale(), but that is used as an example of this new capability.
Additionally this will also allow us to get rid of `// +genclient:noStatus` and fix guessing of the "updateStatus" subresource presence based on the existence of '.Status' field.
Basically you will have to add following into all types you want to generate `UpdateStatus()` for:

```
// +genclient:method=UpdateStatus,verb=update,subresource=status
```

This allows further extension of the client without writing an expansion (which proved to be pain to maintain and copy...). Also allows to customize native CRUD methods if needed (input/output types).

```release-note
NONE
```

### CODE REVIEW

**Comment by mfojtik** [comment]
@deads2k @sttts i will try to fix failing tests tomorrow and check if I captured all "scalable" resource.

I don't want to start picking up the new methods in this PR because it will explode in size, but I'm happy to send a follow ups adopting these in controllers/etc. WDYT?

File: federation/client/clientset_generated/federation_clientset/typed/extensions/v1beta1/daemonset.go [hunk 1]
```diff
@@ -42,6 +42,9 @@ type DaemonSetInterface interface {
 	List(opts v1.ListOptions) (*v1beta1.DaemonSetList, error)
 	Watch(opts v1.ListOptions) (watch.Interface, error)
 	Patch(name string, pt types.PatchType, data []byte, subresources ...string) (result *v1beta1.DaemonSet, err error)
+	GetScale(daemonSetName string, options v1.GetOptions) (*v1beta1.Scale, error)
```
Comment (deads2k) [inline federation/client/clientset_generated/federation_clientset/typed/extensions/v1beta1/daemonset.go]: @DirectXMan12 how are daemonsets scalable?

File: pkg/apis/extensions/v1beta1/zz_generated.conversion.go [hunk 2]
```diff
@@ -325,7 +324,8 @@ func Convert_extensions_DaemonSetList_To_v1beta1_DaemonSetList(in *extensions.Da
 
 func autoConvert_v1beta1_DaemonSetSpec_To_extensions_DaemonSetSpec(in *v1beta1.DaemonSetSpec, out *extensions.DaemonSetSpec, s conversion.Scope) error {
 	out.Selector = (*v1.LabelSelector)(unsafe.Pointer(in.Selector))
-	if err := api_v1.Convert_v1_PodTemplateSpec_To_api_PodTemplateSpec(&in.Template, &out.Template, s); err != nil {
+	// TODO: Inefficient conversion - can we improve it?
```
Comment (mfojtik) [inline pkg/apis/extensions/v1beta1/zz_generated.conversion.go]: not sure how this happened.

File: federation/client/clientset_generated/federation_clientset/typed/extensions/v1beta1/daemonset.go (hunk 1 above)
Comment (mfojtik) [reply federation/client/clientset_generated/federation_clientset/typed/extensions/v1beta1/daemonset.go]: @deads2k hups, I don't think they are...

**Comment by mfojtik** [comment]
@deads2k @sttts hrm. it seems like some API's have custom Scale type and some don't... Not sure if that is a problem or not.

File: staging/src/k8s.io/api/apps/v1beta2/types.go [hunk 3]
```diff
@@ -77,6 +77,8 @@ type Scale struct {
 }
 
 // +genclient
+// +genclient:method=GetScale,verb=get,subresource=scale,result=Scale
```
Comment (deads2k) [inline staging/src/k8s.io/api/apps/v1beta2/types.go]: This looks very wrong.  Having scale defined locally like this means that either someone tweaked the hpa controller (very naughty) or it doesn't work with hpa, right?

File: staging/src/k8s.io/api/core/v1/types.go [hunk 4]
```diff
@@ -2951,6 +2951,8 @@ type ReplicationControllerCondition struct {
 }
 
 // +genclient
+// +genclient:method=GetScale,verb=get,subresource=scale,result=k8s.io/api/extensions/v1beta1.Scale
```
Comment (deads2k) [inline staging/src/k8s.io/api/core/v1/types.go]: yay

File: staging/src/k8s.io/api/apps/v1beta2/types.go (hunk 3 above)
Comment (mfojtik) [reply staging/src/k8s.io/api/apps/v1beta2/types.go]: @deads2k yeah, I wondered why apps.Scale exists... Thought there is just one Scale that should be used everywhere.

File: federation/client/clientset_generated/federation_clientset/typed/extensions/v1beta1/fake/fake_deployment.go [hunk 5]
```diff
@@ -136,3 +136,25 @@ func (c *FakeDeployments) Patch(name string, pt types.PatchType, data []byte, su
 	}
 	return obj.(*v1beta1.Deployment), err
 }
+
+// GetScale takes name of the deployment, and returns the corresponding deployment object, and an error if there is any.
```
Comment (sttts) [inline federation/client/clientset_generated/federation_clientset/typed/extensions/v1beta1/fake/fake_deployment.go]: wrong comment

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go [hunk 6]
```diff
@@ -193,9 +200,88 @@ func (g *genFakeForType) GenerateType(c *generator.Context, t *types.Type, w io.
...
+		if e.HasVerb("patch") {
+			sw.Do(adjustTemplate(e.VerbName, e.VerbType, patchTemplate), m)
+		}
+	}
+
 	return sw.Error()
 }
 
+// adjustTemplate adjust the origin verb template using the expansion name.
+// TODO: Make the verbs in templates parametrized so the strings.Replace() is
```
Comment (sttts) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go]: yes please

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go [hunk 7]
```diff
@@ -193,9 +200,88 @@ func (g *genFakeForType) GenerateType(c *generator.Context, t *types.Type, w io.
 		sw.Do(patchTemplate, m)
 	}
 
+	// generate expansion methods
```
Comment (sttts) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go]: update comment

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go (hunk 6 above)
Comment (mfojtik) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go]: will be a follow up

File: staging/src/k8s.io/api/apps/v1beta2/types.go (hunk 3 above)
Comment (mfojtik) [reply staging/src/k8s.io/api/apps/v1beta2/types.go]: @DirectXMan12 @deads2k I check the `podautoscaler/horizontal.go` and what it does is `scale = Get() -> ext.Scale`  and then `Update(kind, scale)`...

It seems like the Scale struct in this group was copied by mistake and it is not used anywhere? Also I guess we should have just one Scale (looks like autoscaling group would be a good destination for this). WDYT?

Comment (mfojtik) [reply staging/src/k8s.io/api/apps/v1beta2/types.go]: @deads2k i don't think this is a blocker for this PR, once we start plumbing the code to use this new client methods, the first thing we have to do is reconcile the `Scale` usage in kube and regenerate the client to pickup whatever destination we pick for it. Agree?

Comment (deads2k) [reply staging/src/k8s.io/api/apps/v1beta2/types.go]: > @deads2k i don't think this is a blocker for this PR, once we start plumbing the code to use this new client methods, the first thing we have to do is reconcile the Scale usage in kube and regenerate the client to pickup whatever destination we pick for it. Agree?

It's not a blocker.  I don't want you to change types here.  I think someone was confused.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_type.go [hunk 8]
```diff
@@ -77,20 +77,71 @@ func (g *genClientForType) GenerateType(c *generator.Context, t *types.Type, w i
...
+		"DeleteOptions":             c.Universe.Type(types.Name{Package: "k8s.io/apimachinery/pkg/apis/meta/v1", Name: "DeleteOptions"}),
+		"ListOptions":               c.Universe.Type(types.Name{Package: "k8s.io/apimachinery/pkg/apis/meta/v1", Name: "ListOptions"}),
+		"GetOptions":                c.Universe.Type(types.Name{Package: "k8s.io/apimachinery/pkg/apis/meta/v1", Name: "GetOptions"}),
+		"PatchType":                 c.Universe.Type(types.Name{Package: "k8s.io/apimachinery/pkg/types", Name: "PatchType"}),
+		"watchInterface":            c.Universe.Type(types.Name{Package: "k8s.io/apimachinery/pkg/watch", Name: "Interface"}),
+		"RESTClientInterface":       c.Universe.Type(types.Name{Package: "k8s.io/client-go/rest", Name: "Interface"}),
+		"schemeParameterCodec":      c.Universe.Variable(types.Name{Package: filepath.Join(g.clientsetPackage, "scheme"), Name: "ParameterCodec"}),
+		"GroupVersionKind":          c.Universe.Type(types.Name{Package: "k8s.io/apimachinery/pkg/runtime/schema", Name: "GroupVersionKind"}),
+		"UnsafeGuessKindToResource": c.Universe.Function(types.Name{Package: "k8s.io/apimachinery/pkg/api/meta", Name: "UnsafeGuessKindToResource"}),
```
Comment (deads2k) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_type.go]: This still hurts me a bit.  Pre-existing, right?

Comment (mfojtik) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_type.go]: no, this can be nuked... it was there to support the GetScale(kind) thing. good catch!

Comment (mfojtik) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_type.go]: (nuked)

**Comment by deads2k** [comment]
@kubernetes/sig-cli-feature-requests @kubernetes/sig-api-machinery-pr-reviews @caesarxuchao this gives us the ability to generate clients for Scale subresources. I've seen a couple attempts in the cli to standardize. This should help a lot.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go [hunk 9]
```diff
@@ -67,6 +161,8 @@ type Tags struct {
 	// +genclient:skipVerbs=get,update
 	// +genclient:onlyVerbs=create,delete
 	SkipVerbs []string
+	// +genclient:method=Instantiate,verb=create,subresource=instantiate,input=DeploymentRequest,result=DeploymentConfig
```
Comment (deads2k) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go]: This example should be for scale.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go [hunk 10]
```diff
@@ -54,6 +55,99 @@ var ReadonlyVerbs = []string{
...
+}
+
+// Extensions allows to extend the default set of client verbs
+// (CRUD+watch+patch+list+deleteCollection) for a given type with custom defined
+// verbs. Custom verbs can have custom input and result types and also allow to
+// use a sub-resource in a request instead of top-level resource type.
+//
+// Example:
+//
+// +genclient:method=Instantiate,verb=create,subresource=instantiate,input=DeploymentRequest,result=DeploymentConfig
```
Comment (deads2k) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go]: This example should be for scale.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go [hunk 11]
```diff
@@ -146,16 +242,96 @@ func ParseClientGenTags(lines []string) (Tags, error) {
...
+			continue
+		}
+		for _, value := range values {
+			// the value comes in this form: "Foo,verb=create"
+			ext := extension{}
+			parts := strings.Split(value, ",")
+			if len(parts) == 0 {
+				return nil, fmt.Errorf("invalid of empty extension verb name: %q", value)
+			}
+			// The first part represents the name of the extension
```
Comment (deads2k) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go]: I don't like forcing order of sub elements. Issue to make this more flexible please.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go (hunk 10 above)
Comment (mfojtik) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go]: all fixed.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go (hunk 9 above)
Comment (mfojtik) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go]: fixed.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go [hunk 12]
```diff
@@ -146,16 +242,96 @@ func ParseClientGenTags(lines []string) (Tags, error) {
...
+			ext.VerbName = parts[0]
+			if len(ext.VerbName) == 0 {
+				return nil, fmt.Errorf("must specify a verb name (// +genclient:method=Foo,verb=create)")
+			}
+			// Other parts are additional configuration (verb=TYPE is required)
+			params := parts[1:]
+			for _, p := range params {
+				parts := strings.Split(p, "=")
```
Comment (deads2k) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go]: This looks like a helper for parsing a token.  Nothing else needed this?

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go [hunk 13]
```diff
@@ -146,16 +242,96 @@ func ParseClientGenTags(lines []string) (Tags, error) {
...
+				key, val := strings.TrimSpace(parts[0]), strings.TrimSpace(parts[1])
+				if len(val) == 0 {
+					return nil, fmt.Errorf("empty value of %q for %q extension", key, ext.VerbName)
+				}
+				switch key {
+				case "verb":
+					ext.VerbType = val
+				case "subresource":
+					ext.SubResourcePath = val
+				case "input":
```
Comment (deads2k) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go]: I would separate my parse (gather the values), from my validate (are they supported), to eliminate ordering dependencies in the tag.

**Comment by deads2k** [comment]
Some comments on the tag parsing.  All generators tend to be verbose, so the rest lgtm.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go (hunk 11 above)
Comment (mfojtik) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go]: the //+genclient:method=NAME is the required token we looking for, the ordering of the following arguments should not be forced (as you noted in comment bellow).

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go (hunk 12 above)
Comment (mfojtik) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go]: there is no other generator command that takes extra arguments (onlyVerbs, skipVerbs are all main tokens), so nothing in generator AFAIK.

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go (hunk 13 above)
Comment (mfojtik) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/util/tags.go]: fixed, order is no longer required and the validation moved below the loop.

**Comment by mfojtik** [comment]
@deads2k thanks, comments addressed.

**Comment by mfojtik** [comment]
Documentation: https://github.com/kubernetes/community/pull/1008

**Comment by DirectXMan12** [comment]
I'm not entirely certain what the intention of this is -- it doesn't really help with the problems we have with scale, AFAICT.  Sure, the GetScale and UpdateScale methods *might* be convenient, but they don't really represent the fundamental purpose of the scale endpoint, which is to be completely agnostic to underlying resource.  A true scale client would *only* know about scale, and accept arbitrary group-resources.

If we're dropping ScaleExpansion in favor of something else, that's fine, but I don't see how this helps us.

**Comment by mfojtik** [comment]
@DirectXMan12 description updated. This PR has no intention to fix the Scale or refactor it. GetScale() and UpdateScale() are demonstration of how this can be used (actually these methods can be picked up by the CLI for now).

**Comment by deads2k** [comment]
> I'm not entirely certain what the intention of this is -- it doesn't really help with the problems we have with scale, AFAICT. Sure, the GetScale and UpdateScale methods might be convenient, but they don't really represent the fundamental purpose of the scale endpoint, which is to be completely agnostic to underlying resource. A true scale client would only know about scale, and accept arbitrary group-resources.

Spoke in slack.  This provides a client that can be used to call the scale subresource for things like our CLI the way it is structured today. We *also* want a generic scale client, but the two are complementary, not exclusionary.

**Comment by lavalamp** [comment]
I think an issue or a design doc is probably a good idea for this sort of change.

File: pkg/api/types.go [hunk 14]
```diff
@@ -2607,6 +2607,8 @@ type ReplicationControllerCondition struct {
 }
 
 // +genclient
+// +genclient:method=GetScale,verb=get,subresource=scale,result=k8s.io/kubernetes/pkg/apis/extensions.Scale
```
Comment (caesarxuchao) [inline pkg/api/types.go]: Why not treating the Scale subresource the same way we treat Status subresource? We defined templates for UpdateStatus and optionally generated the `UpdateStatus()` methods.

Are there other use cases for the `method` UI? It's useful to mass generating regular CRUD operations for subresources, but for that use case, the approach we took for "Status" is easier to understand.

Comment (caesarxuchao) [reply pkg/api/types.go]: @mfojtik btw, it would be great if you can update the client-gen [doc](https://github.com/kubernetes/community/blob/master/contributors/devel/generating-clientset.md) for the last major improvement you did for client-gen :)

Comment (liggitt) [reply pkg/api/types.go]: > the approach we took for "Status" is easier to understand.

That approach doesn't work for scale. The type passed to scale is not the same as the main resource, and the field name that indicates you can scale varies

Comment (sttts) [reply pkg/api/types.go]: @caesarxuchao about use-cases for `method`: this is for thirdparty apiservers and their api groups. They might have more (and different) subresources than we have in kube with mainly `/status` and `/scale`. This PR enables them to use client-gen effectively. I would prefer a general method like the proposed one than people re-implementing plenty of expansion methods. The former are declarative and stable, the later will break if we change an interface somewhere, e.g. the rest client.

Comment (mfojtik) [reply pkg/api/types.go]: @caesarxuchao i don't think the way we generate Status is good TBH. We rely on the presence of `.Status` field and if that field exists we generate the `UpdateStatus()`, but in case the `.Status` field exists but the `status` subresource don't you need to indicate that with tag...  tbh, I don't like that sort of magic :-)

I believe that using `// +genclient:method=UpdateStatus,subresource=status` is more explicit. Not super-easy to use but leads to cleaner generation :-) WDYT? // cc @sttts

Comment (sttts) [reply pkg/api/types.go]: 100% :+1: for @mfojtik's argument with "that sort of magic" regarding the `/status` subresource.

Comment (caesarxuchao) [reply pkg/api/types.go]: Thanks for the explanation. The thirdparty use cases and explicit tags for `UpdateStatus` sound good.
